fairscale/nn/data_parallel/gossip/gossiper.py

```diff
@@ -0,0 +1,287 @@
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """ Single gossip step """
+        raise NotImplementedError
+
+
+class PushSum(Gossiper):
+    """ 1-peer Push-Sum consensus averaging module """
+
+    def mix(
+        self, out_msg: torch.Tensor, ps_weight: torch.Tensor, residual: bool = False
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """ Consensus averaging step """
+        # out_msg must be on the correct device
+        assert out_msg.device.type == self.device.type
+        if self.logger is not None:
+            self.logger.debug("in/out -peers {}/{}".format(self.in_edges, self.out_edges))
+
+        # prepare messages for gossip
+        mixed_out_msgs = self.mix_out_msg_(out_msg, ps_weight, residual)
+
+        # non-blocking send
+        for out_edge in self.out_edges:
+            msg = next(mixed_out_msgs)
+            assert self.rank == out_edge.src
+            req = dist.broadcast(tensor=msg, src=out_edge.src, group=out_edge.process_group, async_op=True,)
+            self.out_msg_buffer.append((req, msg))
+
+        # blocking recv w/ some code optimization to avoid buffer prep overhead
+        if len(self.in_edges) == 1 and residual:
+            in_edge = self.in_edges[0]
+            dist.broadcast(tensor=self.in_msg_buffer, src=in_edge.src, group=in_edge.process_group)
+
+        # regular non-blocking recv
+        else:
+            # prepare in-msg buffer
+            if not residual:
+                self.in_msg_buffer.copy_(next(mixed_out_msgs))
+            else:
+                self.in_msg_buffer.zero_()
+
+            for in_edge in self.in_edges:
+                dist.broadcast(
+                    tensor=self.placeholder, src=in_edge.src, group=in_edge.process_group,
+                )
+                self.in_msg_buffer.add_(self.placeholder)  # type: ignore
+
+        self.refresh_peers_()
+        self.clean_msg_buffers_()
+        return self.parse_in_msg_buffer(residual)
+
+
+class PushPull(Gossiper):
+    """ Doubly-stochastic consensus averaging module """
+
+    def mix(
+        self, out_msg: torch.Tensor, ps_weight: torch.Tensor, residual: bool = False
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # out_msg must be on the correct device
+        assert out_msg.device.type == self.device.type
+        if self.logger is not None:
+            self.logger.debug("in/out -peers {}/{}".format(self.in_edges, self.out_edges))
+
+        # prepare messages for gossip
+        mixed_out_msgs = self.mix_out_msg_(out_msg, ps_weight, residual)
+
+        # send-recv w/ some code optimization to avoid buffer prep overhead
+        if len(self.in_edges) == 1 and len(self.out_edges) == 1 and residual:
+            out_edge, in_edge = self.out_edges[0], self.in_edges[0]
+            msg = next(mixed_out_msgs)
+            if not self.passive:
+                dist.broadcast(tensor=msg, src=out_edge.src, group=out_edge.process_group)
```

Comment:
this will be blocking, any benefit in async ? Asking just in case, may or may not make sense

Reply:
cc @blefaudeux, I've added TensorPipe in one of the potential steps for the future. Do you think it would be okay to resolve this comment?

Reply:
yes definitely, thanks for the follow up in the notes ! by the way NCCL is very slow for broadcasts with the default settings (related to env variables), and even not that fast when they are tuned, FYI (recently ported to my attention by the gloo2 tests, which show a much better speed)